File: lc_soln/1772_sort_features_by_popularity/solution.py
from typing import List


class Solution:
    def sort_features(self, features: List[str], responses: List[str]) -> List[str]:
        popularity = collections.Counter(features)
        for response in responses:
            for word in set(response.split()):
                popularity[word] += 1
        return sorted(features, key=lambda f: popularity[f], reverse=True)

File: lc_soln/1772_sort_features_by_popularity/solution.cpp
class Solution {
public:
    vector<string> sortFeatures(vector<string> &features, vector<string> &responses) {
        unordered_map<string, int> popularity;
        for (const string& feature : features) {
            popularity[feature]++;
        }

        for (const string& response : responses) {
            unordered_set<string> seen;
            istringstream iss(response);
            string token;
            while (iss >> token) {
                if (!token.empty() && seen.count(token) == 0) {
                    popularity[token]++;
                    seen.insert(token);
                }
            }
        }
        stable_sort(features.begin(), features.end(), [popularity](const string& a, const string& b) {
            return popularity.at(a) > popularity.at(b);
        });
        return features;
    }
};
